Cluster notes

benarmston opened the pull request:
This PR implements #232.  Currently, a cluster can have `engineering` and `customer` notes.  Admin users can view, create and update both kinds of notes, site contacts can view, create and update only `customer` notes.

@bobwhitelock I have a few questions:

 1. Is this along the lines of what you were expecting?
 2. I'm assuming that we want to add the ability for the `customer` notes to be edited created through the non-RailsAdmin UI.  Is that correct?
 3. Do we use the RailsAdmin UI in production for creating/editing/deleting admin content?
 4. Any other comments you have would be appreciated.

bobwhitelock commented on app/decorators/note_decorator.rb:
```diff
@@ -0,0 +1,21 @@
+    if current_user.admin?
+      "No #{flavour} notes have been addded yet."
+    else
+      "No notes have been addded yet."
```
FYI `s/addded/added/` here and above :slightly_smiling_face:

bobwhitelock commented on app/models/concerns/admin_config/note.rb:
```diff
@@ -0,0 +1,20 @@
+
+        configure :flavour do
+          help <<~EOF
+            Valid flavours are #{Note::FLAVOURS.join(', ')}.
```
FWIW if you define a `Note#flavour_enum` method this will be automatically picked up by `rails-admin` and only these flavours will be selectable, then you won't need this help :slightly_smiling_face:. See here for where we do something similar: https://github.com/alces-software/alces-flight-center/blob/a3c31dadf894ea044def036e5809901a0bbf6f6c/app/models/cluster.rb#L39-L43

benarmston commented:
@bobwhitelock thanks for the comments above.  I've addressed the issues you mentioned, implemented, create/update support and written some tests.  Additional features that we're likely going to want at some point might include:

 - Preview support, similar to the github comment box.
 - A rich text editor, similar to the github comment box.

I can look at implementing one or both of those, or move onto something else.

bobwhitelock commented on app/controllers/notes_controller.rb:
```diff
@@ -0,0 +1,56 @@
+      flash[:success] = 'Notes updated'
+      redirect_to @note.decorate.path
+    else
+      flash[:error] = "Your notes were not updated: #{format_errors(@note)}"
```
If I create some notes, then edit them and attempt to delete all content, I get an error as `Your notes were not updated: description can't be blank` - however I can imagine valid use cases where a User may want to do this, e.g.:

1. a User adds a note to a Cluster, which previously had no notes, about some temporary issue to be aware of related to the Cluster;

2. later this issue no longer exists so they want to delete everything in the note;

3. currently to do this they would need to still enter some placeholder text so that the existing note can be saved.

Possibly we should either allow `description` to be empty when updating, or `destroy` the existing note when `update` is submitted with a blank description (this would also have the advantage of re-displaying the 'no notes yet' message in this situation). What do you think to this?

benarmston replied:
Let's destroy the note when it has no description.

bobwhitelock commented on app/controllers/notes_controller.rb:
```diff
@@ -0,0 +1,56 @@
+  private
+
+  def note_from_params
+    cluster = cluster_from_params
```
We have some code in `ApplicationController` to find the top-level object that the User is on the dashboard for (which we call the scope): https://github.com/alces-software/alces-flight-center/blob/ee18cddfdbd57440550b80b51284cc21e2516405/app/controllers/application_controller.rb#L69-L88

So I think you could make use of this by using `@cluster` here, which would avoid `find`ing the Cluster twice.

(I realise this is not that discoverable atm, and there's probably some room for improvement with this stuff!)

benarmston replied:
OK. I'll swap out to using the auto assigned `@cluster`.  Thanks for pointing me at this.

Regarding discoverability, I think that's going to be an issue for anyone joining any largeish project.

bobwhitelock commented on app/views/notes/new.html.erb:
```diff
@@ -0,0 +1,17 @@
+            </div>
+        <% end %>
+    </div>
+<% end %>
```
The `new` and `edit` views currently look very similar - possibly worth extracting a partial for these which could just be passed the text differences? Would avoid them getting out-of-sync and/or us needing to remember to update both together.

benarmston replied:
OK. I'll do that.

bobwhitelock commented on app/views/notes/new.html.erb:
```diff
@@ -0,0 +1,17 @@
+        </p>
+        <%= form_for note.form_path do |f| %>
+            <div class="form-group">
+                <%= f.text_area :description, class: 'form-control', rows: 10 %>
```
Possibly worth indicating that this description will be rendered as markdown, as customers may not be aware of this (same for the `edit` form).

benarmston replied:
I'll add a very simple indication of that.  I'll leave link to the markdown flavour we support for a future enhancement.

benarmston commented on app/controllers/notes_controller.rb:
```diff
@@ -0,0 +1,56 @@
+      render :show
+    else
+      render :new
+    end
```
I've not much preference.  I'll change it as you suggest.

benarmston commented:
@bobwhitelock I've addressed all of your comments.

bobwhitelock commented on app/controllers/notes_controller.rb:
```diff
@@ -0,0 +1,51 @@
+    if note_params[:description].blank?
+      @note.destroy
+      flash[:success] = 'Notes removed'
+      redirect_to cluster_path(@note.cluster)
```
Do we not want to `redirect_to @note.decorate.path` in this case too? Being redirected back to the cluster overview page seems slightly surprising.

benarmston replied:
I don't think so.  If we did, we'd have the following interaction:

 - user: visits the note page
 - user: decides to remove note
 - centre: informs user (via flash) that the removal was successful
 - centre: asks user if they would like to create the note that they just deleted.

I can't imagine the use case in which that would leave the user on the correct page.

bobwhitelock replied:
I guess the same is true with the language suggesting 'No notes have been added for this cluster yet' if they then navigate back to the notes page, which seems surprising in the same way since notes have been added (and then later removed) for the cluster - so possibly the wording should be changed to not imply this, to something like 'there are currently no $flavour notes for this cluster'.

I was thinking that, from a users perspective, their intention when deleting all the content in an existing note is to just have that note be empty, and the fact that behind the scenes this actually deletes the model is somewhat an implementation detail. Therefore it could surprise people that if they leave a single character in the description they will be redirected back to the same page, but if they delete one more character they will be redirected to a different page, even though the original page is still a valid URL they can visit (and they can immediately navigate back here if they want).

In any case maybe real Users will think differently to me, so happy for you to tweak the wording/change the redirect and/or just merge this as you prefer, we can always adjust in response to user feedback :slightly_smiling_face:.

benarmston replied:
I like the changed wording. I'll make that change.

benarmston replied:
If I understand your last paragraph correctly, you are happy for me to merge this PR to develop.  Is that correct?

bobwhitelock replied:
Yep, fine by me :slightly_smiling_face:
